File: 강병호/6week/[1932] 정수 삼각형.py
'''
* Author    : BYEONGHO KANG
* Date      : 2024.11.8(Fri)
* Runtime   : 120 ms
* Memory    : 40264 KB
* Algorithm : 다이나믹 프로그래밍
'''


import sys
input = sys.stdin.read

data = input().splitlines()
n = int(data[0])
triangle = [list(map(int, line.split())) for line in data[1:n+1]]

dp = [[0] * (i + 1) for i in range(n)]
dp[0][0] = triangle[0][0] 

for i in range(1, n):
    for j in range(i + 1):
        if j == 0:
            # left
            dp[i][j] = dp[i - 1][j] + triangle[i][j]
        elif j == i:
            # right
            dp[i][j] = dp[i - 1][j - 1] + triangle[i][j]
        else:
            # middle
            dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i][j]

print(max(dp[n - 1]))

File: 이정훈/6주차/[1932] 정수 삼각형.cpp
/*
* Author    : Lee JeongHun
* Date      : 2024.11.09(Sat)
* Runtime   : 36 ms
* Memory    : 3076 KB
* Algorithm : DP
*/

#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

int main()
{
    // 입력 처리
    int n;
    cin >> n;

    vector<vector<int>> triangle(n);

    for(int i = 0; i < n; ++i)
    {
        triangle[i].reserve(i);

        for(int j = 0; j <= i; ++j)
        {
            int input;
            cin >> input;
            triangle[i].push_back(input);
        }
    }

    // 점화식
    // j = 0 or i인 경우에는 경로가 하나밖에 존재하지 않습니다.
    // 그 외에는 result[i][j] = max(result[i - 1][j - 1], result[i - 1][j]) + triangle[i][j]입니다.

    for(int i = 1; i < n; ++i)
    {
        for(int j = 0; j <= i; ++j)
        {
            if(j == 0)
            {
                triangle[i][j] += triangle[i - 1][0];
            }
            else if(j == i)
            {
                triangle[i][j] += triangle[i - 1][i - 1];
            }
            else
            {
                triangle[i][j] += max(triangle[i - 1][j - 1], triangle[i - 1][j]);
            }
        }
    }

    // 정답 출력
    int answer = *max_element(triangle[n - 1].begin(), triangle[n - 1].end());

    cout << answer;

    return 0;
}